File: src/main/java/com/matchFit/common/aop/TimingAspect.java
package com.matchFit.common.aop;

import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import lombok.RequiredArgsConstructor;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;

@Aspect
@Component
@RequiredArgsConstructor
public class TimingAspect {

    private final MeterRegistry meterRegistry;

    @Around("execution(public * com.matchFit..service..*(..))")
    public Object timeServiceMethods(ProceedingJoinPoint pjp) throws Throwable {
        MethodSignature signature = (MethodSignature) pjp.getSignature();
        String className = signature.getDeclaringType().getSimpleName();
        String methodName = signature.getName();

        Timer.Sample sample = Timer.start(meterRegistry);
        try {
            return pjp.proceed();
        } finally {
            Timer timer = Timer.builder("method.execution")
                    .description("Execution time of methods (AOP)")
                    .tags("class", className, "method", methodName)
                    .publishPercentileHistogram()
                    .register(meterRegistry);
            sample.stop(timer);
        }
    }
}

File: src/main/kotlin/com/matchFit/common/aop/TimingAspect.kt
package com.matchFit.common.aop

import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Timer
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.stereotype.Component

@Aspect
@Component
class TimingAspect(private val meterRegistry: MeterRegistry) {
    @Around("execution(public * com.matchFit..service..*(..))")
    @Throws(Throwable::class)
    fun timeServiceMethods(pjp: ProceedingJoinPoint): Any? {
        val signature = pjp.signature as MethodSignature
        val className = signature.declaringType.simpleName
        val methodName = signature.name

        val sample = Timer.start(meterRegistry)
        try {
            return pjp.proceed()
        } finally {
            val timer = Timer.builder("method.execution")
                .description("Execution time of methods (AOP)")
                .tags("class", className, "method", methodName)
                .publishPercentileHistogram()
                .register(meterRegistry)
            sample.stop(timer)
        }
    }
}
